w3m-hist.el

```diff
@@ -718,6 +718,116 @@ are accessible by PREV and NEXT operations."
     (setq w3m-history-flat (w3m-history-slimmed-history-flat))
     (w3m-history-tree position)))
 
+(defvar w3m-scrub-command-list
+  '(("/usr/bin/shred" "-fuz") ("/bin/rm" "-f"))
+  "List of methods to use for attempting to securely delete temporary files.
+
+Function `w3m-scrub-command' uses the first entry on the list
+that it finds available on the host OS; if none are found, it
+will use the insecure emacs function `delete-file'.")
+
+(defun w3m-history-scrub ()
+  "Initializes history, cache, cookies, and temp files.
+
+The method used to delete files is configurable to your personal
+level of paranoia, via `w3m-scrub-command-list'.
+
+Note that some operations performed by this function may impact
+operation of the external w3m browser. Also, be advised that
+other emacs features outside the scope of `emacs-w3m' may
+independantly be saving varibles, for example in order to restore
+a crashed emacs session."
+  ; TODO:
+  ; + Use a defcustom or other user-friendly mechanism to allow
+  ;   fine-grained customization of what to scrub.
+  ; + When invoked with a prefix arg, allow the scrub to be limited to a
+  ;   timeframe, for elements that store timestamp information.
+  (interactive)
+  (let* ((w3m-fb-mode nil)
+         (w3m-message-silent t)
+         (stdout-buf (generate-new-buffer "*w3m-scrub-history*"))
+         (cookie-buf (get-buffer " *w3m-cookie-parse-temp*"))
+         (bufs (w3m-list-buffers t))
+         (cmds w3m-scrub-command-list)
+         (cmd (progn
+                (while (and (setq cmd (pop cmds))
+                            (not (file-executable-p (car cmd)))))
+                cmd))
+         (cmd-string (if cmd
+                       (format "%s %s" (car cmd) (cdr cmd))
+                      "emacs' delete-file"))
+         (file "")
+         (files (append
+                  (list w3m-arrived-file w3m-cookie-file)
+                  w3m-form-textarea-post-files
+                  w3m-form-textarea-files
+                  (directory-files
+                    (file-chase-links (expand-file-name w3m-form-textarea-directory))
+                    'full "[^.]" 'nosort)
+                  (directory-files w3m-profile-directory 'full
+                    "^w3m\\(cache\\|cookie\\|el\\|src\\|tmp\\)" 'nosort))))
+   (switch-to-buffer stdout-buf)
+   (insert "Beginning emacs-w3m history scrub "
+           (format-time-string "%Y-%m-%d %H:%M")
+           "\n\nDeleting files ...\n")
+   (if cmd
+     (while files
```

Comment:
dolist() may be more readable in this case.  CL macros are allowed to use in emacs-w3m.

Reply:
Thanks. Done, in 0c6f5ea